src/pretix/base/datasync/datasync.py

```diff
@@ -0,0 +1,425 @@
+    )
+    sync_provider = models.CharField(blank=False, null=False, max_length=128)
+    order_position = models.ForeignKey(
+        OrderPosition, on_delete=models.CASCADE, related_name="synced_objects", blank=True, null=True,
+    )
+    external_object_type = models.CharField(blank=False, null=False, max_length=128)
+    external_pk_name = models.CharField(blank=False, null=False, max_length=128)
+    external_pk_value = models.CharField(blank=False, null=False, max_length=128)
+    external_link_href = models.CharField(blank=True, null=True, max_length=255)
+    external_link_display_name = models.CharField(blank=True, null=True, max_length=255)
+    timestamp = models.DateTimeField(blank=False, null=False, auto_now_add=True)
+
+    def external_link_html(self):
+        if not self.external_link_display_name:
+            return None
+
+        prov, meta = sync_targets.get(identifier=self.sync_provider)
+        if prov:
+            return prov.get_external_link_html(self.order.event, self.external_link_href, self.external_link_display_name)
+
+
+@receiver(periodic_task, dispatch_uid="data_sync_periodic")
+def on_periodic_task(sender, **kwargs):
+    sync_all.apply_async()
+
+
+sync_targets = EventPluginRegistry({"identifier": lambda o: o.identifier})
+
+
+def sync_event_to_target(event, target_cls, queued_orders):
+    with scope(organizer=event.organizer):
+        with target_cls(event=event) as p:
+            # TODO: should I somehow lock the queued orders or events, to avoid syncing them twice at the same time?
+            p.sync_queued_orders(queued_orders)
+
+
+@app.task()
+def sync_all():
+    with scopes_disabled():
+        queue = (
+            OrderSyncQueue.objects
+            .select_related("order")
+            .prefetch_related("order__event")
+            .filter(Q(not_before__isnull=True) | Q(not_before__lt=now()))[:1000]
+        )
+        grouped = groupby(sorted(queue, key=lambda q: (q.sync_provider, q.order.event.pk)), lambda q: (q.sync_provider, q.order.event))
+        for (target, event), queued_orders in grouped:
+            target_cls, meta = sync_targets.get(identifier=target, active_in=event)
+            # TODO: what should i do if the sync plugin got deactivated in the meantime?
+            sync_event_to_target(event, target_cls, queued_orders)
+
+
+class SyncConfigError(Exception):
+    def __init__(self, messages, full_message=None):
+        self.messages = messages
+        self.full_message = full_message
+
+
+StaticMapping = namedtuple('StaticMapping', ('pk', 'pretix_model', 'external_object_type', 'pretix_pk', 'external_pk', 'property_mapping'))
+
+
+class OutboundSyncProvider:
+    max_attempts = 5
+
+    def __init__(self, event):
+        self.event = event
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    @classmethod
+    @property
+    def display_name(cls):
+        return str(cls.identifier)
+
+    """
```

Comment:
Unlike Java, Python docstrings live within the function they describe, not above it: https://peps.python.org/pep-0257/#multi-line-docstrings

Reply:
that was the reason why sphinx didn't want to output documentation for my classes :)